conn_js.go

```diff
@@ -0,0 +1,196 @@
+
+const (
+	webSocketStateConnecting = 0
+	webSocketStateOpen       = 1
+	webSocketStateClosing    = 2
+	webSocketStateClosed     = 3
+)
+
+var errIsClosed = errors.New("connection is closed")
+
+// Conn implements net.Conn interface for WebSockets in js/wasm.
+type Conn struct {
+	js.Value
+	messageHandler *js.Func
+	closeHandler   *js.Func
+	mut            sync.Mutex
+	incomingData   chan []byte
+	currData       *bytes.Buffer
+	closeSignal    chan struct{}
+}
+
+func (c *Conn) Read(b []byte) (int, error) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	if err := c.checkOpen(); err != nil {
+		return 0, io.EOF
+	}
+
+	if c.currData == nil {
+		select {
+		case data := <-c.incomingData:
+			c.currData = bytes.NewBuffer(data)
+		case <-c.closeSignal:
+			return 0, io.EOF
+		}
+	}
+
+	n, err := c.currData.Read(b)
+	if err == io.EOF {
+		c.currData = nil
+		return n, nil
+	}
+	return n, err
+}
+
+func (c *Conn) checkOpen() error {
+	state := c.Get("readyState").Int()
+	switch state {
+	case webSocketStateClosed, webSocketStateClosing:
+		return errIsClosed
+	}
+	return nil
+}
+
+func (c *Conn) Write(b []byte) (n int, err error) {
+	// c.mut.Lock()
+	// defer c.mut.Unlock()
+	if err := c.checkOpen(); err != nil {
+		return 0, err
+	}
+	c.Call("send", string(b))
+	return len(b), nil
+}
+
+// Close closes the connection. Only the first call to Close will receive the
+// close error, subsequent and concurrent calls will return nil.
+// This method is thread-safe.
+func (c *Conn) Close() error {
+	// c.mut.Lock()
+	// defer c.mut.Unlock()
+	c.Call("close")
+	if c.messageHandler != nil {
+		c.Call("removeEventListener", "message", *c.messageHandler)
+		c.messageHandler.Release()
+	}
+	if c.closeHandler != nil {
+		c.Call("removeEventListener", "close", *c.closeHandler)
+		c.closeHandler.Release()
+	}
+	return nil
+}
+
+func (c *Conn) LocalAddr() net.Addr {
+	// TODO(albrow): is there some way to get the local address?
+	return NewAddr("0.0.0.0:0")
+}
+
+func (c *Conn) RemoteAddr() net.Addr {
+	rawURL := c.Get("url").String()
+	withoutPrefix := strings.TrimPrefix(rawURL, "ws://")
+	withoutSuffix := strings.TrimSuffix(withoutPrefix, "/")
+	return NewAddr(withoutSuffix)
+}
+
+func (c *Conn) SetDeadline(t time.Time) error {
+	// Not supported
+	return nil
+}
+
+func (c *Conn) SetReadDeadline(t time.Time) error {
+	// Not supported
+	return nil
+}
+
+func (c *Conn) SetWriteDeadline(t time.Time) error {
+	// Not supported
+	return nil
+}
+
+// NewConn creates a Conn given a regular js/wasm WebSocket Conn.
+func NewConn(raw js.Value) *Conn {
+	conn := &Conn{
+		Value:        raw,
+		incomingData: make(chan []byte),
+		closeSignal:  make(chan struct{}),
+	}
+	conn.setUpHandlers()
+	return conn
+}
+
+func (c *Conn) setUpHandlers() {
+	// c.mut.Lock()
+	// defer c.mut.Unlock()
+	if c.messageHandler != nil {
+		// Message handlers already created. Nothing to do.
+		return
+	}
+	messageHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		go func() {
```

Comment:
This could deliver messages out of order. We need some kind of in-order buffer. We can do that by either:

1. Spinning up a dedicated goroutine to read from this channel.
2. Ideally, just writing directly into a shared buffer (protected by a mutex). Availability of data to read could be signaled by writing an empty `struct{}{}` to a flag channel.

Reply:
@Stebalien take a look at the latest version and let me know what you think. Just need to figure out how and when to reset the buffer.